packages/zoe/src/contracts/barterExchange.js

```diff
@@ -0,0 +1,188 @@
+// @ts-check
+
+import harden from '@agoric/harden';
+import { makeZoeHelpers, defaultAcceptanceMsg } from '../contractSupport';
+
+/** @typedef {import('../zoe').ContractFacet} ContractFacet */
+
+/**
+ * The Barter Exchange ignores the keywords in offers. It takes advantage of
+ * Zoe facet parameters by only paying attention the issuers in proposals.
+ *
+ * The want and give amounts are both treated as minimums. Each successful
+ * trader gets their `want` and may trade with counter-parties who specify any
+ * amount up to their specified `give`.
+ */
+export const makeContract = harden(
+  /** @param {ContractFacet} zcf */ zcf => {
+    // bookOrders is a Map of Maps. The first key is the brand of each offer's
+    // give, and the second key is the brand of their want.
+    // for each offer, we store (see extractOfferDetails) the handle, as well as
+    // keywords, brands, and amount for both `give` and `want`. The keywords are
+    // only used to produce the payout.
+    const bookOrders = new Map();
+    let brandToAmountMath;
+
+    const { rejectOffer, getKeys, assertKeywords } = makeZoeHelpers(zcf);
+
+    assertKeywords(harden(['Asset', 'Price']));
+
+    function lookupBookOrders(brandIn, brandOut) {
+      let ordersMap = bookOrders.get(brandIn);
+      if (!ordersMap) {
+        ordersMap = new Map();
+        bookOrders.set(brandIn, ordersMap);
+      }
+      let ordersArray = ordersMap.get(brandOut);
+      if (!ordersArray) {
+        ordersArray = [];
+        ordersMap.set(brandOut, ordersArray);
+      }
+      return ordersArray;
+    }
+
+    const getKeywordAndBrand = (offerHandle, amountKeywordRecord) => {
+      const keywords = getKeys(amountKeywordRecord);
+      if (keywords.length !== 1) {
+        rejectOffer(
+          offerHandle,
+          `A swap requires giving one type of token for another, ${keywords.length} tokens were provided.`,
+        );
+      }
+      return harden({
+        keyword: keywords[0],
+        brand: Object.values(amountKeywordRecord)[0].brand,
```

Comment:
why not use keyword as the key here? 
```suggestion
        brand: amountKeywordRecord[keyword].brand,
```

or `amountKeywordRecord[keywords[0]].brand`

Reply:
Thanks that's clearer.